src/XMonad/Core.hs

```diff
@@ -169,7 +169,7 @@ newtype ScreenDetail = SD { screenRect :: Rectangle }
 -- instantiated on 'XConf' and 'XState' automatically.
 --
 newtype X a = X (ReaderT XConf (StateT XState IO) a)
```

Comment:
The point of using `deriving via` here is to allow you to inline the contents of these newtypes (and hence obtain the `type role X representational`); the field should be `XConf -> XState -> IO (a, XState)`.